File: flowquery-py/src/parsing/functions/predicate_function.py
"""Base class for predicate functions in FlowQuery."""

from typing import TYPE_CHECKING, Any, Optional

from ..ast_node import ASTNode
from .value_holder import ValueHolder

if TYPE_CHECKING:
    pass


class PredicateFunction(ASTNode):
    """Base class for predicate functions."""

    def __init__(self, name: Optional[str] = None):
        super().__init__()
        self._name = name or self.__class__.__name__
        self._value_holder = ValueHolder()
        self._has_return_expression = True

    @property
    def name(self) -> str:
        return self._name

    @property
    def has_return_expression(self) -> bool:
        return self._has_return_expression

    @has_return_expression.setter
    def has_return_expression(self, value: bool) -> None:
        self._has_return_expression = value

    @property
    def reference(self) -> ASTNode:
        return self.first_child()

    @property
    def array(self) -> ASTNode:
        return self.get_children()[1].first_child()

    @property
    def _return(self) -> Optional[ASTNode]:
        if not self._has_return_expression:
            return None
        return self.get_children()[2]

    @property
    def where(self) -> Optional[ASTNode]:
        if self._has_return_expression:
            if len(self.get_children()) == 4:
                return self.get_children()[3]
        else:
            if len(self.get_children()) == 3:
                return self.get_children()[2]
        return None

    def value(self) -> Any:
        raise NotImplementedError("Method not implemented.")

    def __str__(self) -> str:
        return f"PredicateFunction ({self._name})"

File: src/parsing/functions/predicate_function.ts
import ASTNode from "../ast_node";
import Expression from "../expressions/expression";
import Reference from "../expressions/reference";
import Where from "../operations/where";
import ValueHolder from "./value_holder";

class PredicateFunction extends ASTNode {
    private _name: string;
    protected _valueHolder: ValueHolder = new ValueHolder();
    private _hasReturnExpression: boolean = true;

    constructor(name?: string) {
        super();
        this._name = name || this.constructor.name;
    }

    public get name(): string {
        return this._name;
    }

    public get hasReturnExpression(): boolean {
        return this._hasReturnExpression;
    }

    public set hasReturnExpression(value: boolean) {
        this._hasReturnExpression = value;
    }

    protected get reference(): Reference {
        return this.firstChild() as Reference;
    }

    protected get array(): ASTNode {
        return this.getChildren()[1].firstChild();
    }

    protected get _return(): Expression | null {
        if (!this._hasReturnExpression) return null;
        return this.getChildren()[2] as Expression;
    }

    protected get where(): Where | null {
        if (this._hasReturnExpression) {
            if (this.getChildren().length === 4) {
                return this.getChildren()[3] as Where;
            }
        } else {
            if (this.getChildren().length === 3) {
                return this.getChildren()[2] as Where;
            }
        }
        return null;
    }

    public value(): any {
        throw new Error("Method not implemented.");
    }

    public toString(): string {
        return `PredicateFunction (${this._name})`;
    }
}

export default PredicateFunction;
